network/server/tchannelthrift/node/service.go

```diff
@@ -307,57 +357,75 @@ func (s *service) Write(tctx thrift.Context, req *rpc.WriteRequest) error {
 		s.metrics.write.ReportError(s.nowFn().Sub(callStart))
 		return tterrors.NewBadRequestError(fmt.Errorf("requires datapoint"))
 	}
+
 	dp := req.Datapoint
 	unit, unitErr := convert.ToUnit(dp.TimestampType)
+
 	if unitErr != nil {
 		s.metrics.write.ReportError(s.nowFn().Sub(callStart))
 		return tterrors.NewBadRequestError(unitErr)
 	}
+
 	d, err := unit.Value()
 	if err != nil {
 		s.metrics.write.ReportError(s.nowFn().Sub(callStart))
 		return tterrors.NewBadRequestError(err)
 	}
+
 	timestamp := xtime.FromNormalizedTime(dp.Timestamp, d)
+	nsID, tsID := s.idpool.StringID(req.NameSpace), s.idpool.StringID(req.ID)
 	err = s.db.Write(
-		ctx, ts.StringID(req.NameSpace), ts.StringID(req.ID), timestamp, dp.Value, unit, dp.Annotation)
+		ctx, nsID, tsID, timestamp, dp.Value, unit, dp.Annotation)
+	s.idpool.Release(nsID, tsID)
+
 	if err != nil {
 		s.metrics.write.ReportError(s.nowFn().Sub(callStart))
 		return convert.ToRPCError(err)
 	}
+
 	s.metrics.write.ReportSuccess(s.nowFn().Sub(callStart))
+
 	return nil
 }
 
 func (s *service) WriteBatchRaw(tctx thrift.Context, req *rpc.WriteBatchRawRequest) error {
 	callStart := s.nowFn()
 	ctx := tchannelthrift.Context(tctx)
-	ns := ts.BinaryID(req.NameSpace)
+	nsID := s.idpool.BinaryID(req.NameSpace)
 
 	var errs []*rpc.WriteBatchRawError
+
 	for i, elem := range req.Elements {
 		unit, unitErr := convert.ToUnit(elem.Datapoint.TimestampType)
+
 		if unitErr != nil {
 			errs = append(errs, tterrors.NewBadRequestWriteBatchRawError(i, unitErr))
 			continue
 		}
+
 		d, err := unit.Value()
 		if err != nil {
 			errs = append(errs, tterrors.NewBadRequestWriteBatchRawError(i, err))
 			continue
 		}
+
 		timestamp := xtime.FromNormalizedTime(elem.Datapoint.Timestamp, d)
+		tsID := s.idpool.BinaryID(elem.ID)
 		err = s.db.Write(
-			ctx, ns, ts.BinaryID(elem.ID), timestamp, elem.Datapoint.Value, unit, elem.Datapoint.Annotation)
-		if err != nil {
-			if xerrors.IsInvalidParams(err) {
-				errs = append(errs, tterrors.NewBadRequestWriteBatchRawError(i, err))
-			} else {
-				errs = append(errs, tterrors.NewWriteBatchRawError(i, err))
-			}
+			ctx, nsID, tsID, timestamp, elem.Datapoint.Value, unit, elem.Datapoint.Annotation)
+		s.idpool.Release(tsID)
+
+		switch {
+		case err == nil:
+		case xerrors.IsInvalidParams(err):
+			errs = append(errs, tterrors.NewBadRequestWriteBatchRawError(i, err))
+		default:
+			errs = append(errs, tterrors.NewWriteBatchRawError(i, err))
```

Comment:
This switch is neat, however not super idiomatic.  I'd  be in preference of leaving it as the original writing or to avoid one more level of indentation just:
```go
if err != nil && xerrors.IsInvalidParams(err) { 
  errs = append(errs, tterrors.NewBadRequestWriteBatchRawError(i, err))
} else if err != nil {
  errs = append(errs, tterrors.NewWriteBatchRawError(i, err))
}
```